specs/merkle-tree.md

```diff
@@ -0,0 +1,158 @@
+
+For example, the following Merkle tree would be valid as long as its leaves are submitted on-chain in ascending order by their index. This is true even though the tree was constructed with the order of the leaves reversed.
+```mermaid
+graph LR
+    A[Root] --> B[Hash]
+    A[Root] --> C[Hash]
+    B[Hash] --> D[Hash]
+    B[Hash] --> E[Hash]
+    C[Hash] --> F[Hash]
+    C[Hash] --> G[Hash]
+    D[Hash] --> I["Execute: Chain ID: 5, Index: 3, Data: 0xaa..."]
+    D[Hash] --> J["Execute: Chain ID: 5, Index: 2, Data: 0xbb..."]
+    E[Hash] --> K["Execute: Chain ID: 5, Index: 1, Data: 0xcc..."]
+    E[Hash] --> H["Approve: Chain ID: 5, Index: 0"]
+    F[Hash] --> M["Execute: Chain ID: 420, Index: 3, Data: 0xdd..."]
+    F[Hash] --> N["Execute: Chain ID: 420, Index: 2, Data: 0xee..."]
+    G[Hash] --> O["Execute: Chain ID: 420, Index: 1, Data: 0xff..."]
+    G[Hash] --> L["Approve: Chain ID: 420, Index: 0"]
+```
+
+Sphinx Merkle trees should follow a consistent ordering scheme to ensure they can be executed by a variety of executor implementations without issue. We impose the standard that the leaves should be ordered by index and chain ID ascending.
+
+So for example, if we intended to run some set of transactions on Goerli (5), OP Goerli (420), and Polygon Mumbai (80001). Then the leaves are expected to be ordered with the approval leaf for Goerli first, then all of the transaction leaves on Goerli in ascending order, then the approval leaf for OP Goerli, all the transaction leaves on OP Goerli in ascending order, and so on.
+
+### 3. Every chain ID and index combination must be unique within the tree
+Since the Sphinx Merkle tree leaves contain explicit indexes, it is also possible to construct trees that contain multiple leaves with the same index value and chain ID. However, only one of the leaves would be executable on-chain. Therefore, constructing a tree with multiple leaves with the same chain ID and index would make it ambiguous to the relayer which leaf should be executed, and would result in a tree that cannot be fully executed. So we impose the standard that in a given Sphinx Merkle tree there must be at most one leaf with each chain ID and index combination.
+
+### 4. There must be exactly one approval leaf or cancel leaf per chain
+Since Sphinx Merkle trees can contain an arbitrary set of leaves, it is possible to construct a single tree that contains `APPROVE` or `CANCEL` leaves that approve or cancel multiple separate deployments on a single chain. However, this leads to an ambiguous situation for the relayer where it is unclear which deployment should be approved or canceled.
```

Comment:
Same feedback as the last section: it's confusing to say "it's possible to construct..." in this section because it's not supposed to be possible under this spec

Reply:
The purpose of this paragraph is to explain _why_ the invariant exists by explaining what you could do if it did not and why that would cause issues. This is the main reason why I include so much detail on the possible Merkle trees you could construct which are not valid. I am trying to preemptively answer the auditor's questions regarding potential situations where the Merkle tree may be ambiguous, but still executable.

This was one of Kelvin's main pieces of feedback on the specs in general. He said they do a good job of stating exactly what the code must do, but they do not explain _why_. This prevents the reader from really understanding the purpose of the invariants and stops them from being able to offer higher-level feedback on the overall design of the system. Focusing only on exactly what the system does and not explaining the why forces them to focus only on the low-level details of the specific invariants. I think it would be a mistake to remove this context and I would encourage you to consider adding more reasoning to the invariants in the specs to make sure auditors understand the purpose of the invariants.